Dockerfile

```diff
@@ -2,37 +2,113 @@
+ARG JDK_DIGEST_AMD64="sha256:a16cce6e84fd66cdbf753dd7bc65f6ba2e6d74b907052146709b3dd513dc54b6"
+ARG JDK_DIGEST_ARM64="sha256:f84e436a00db070dd6df8206b3800b8e7799e13a9f75a96581e425707672d193"
 
 # set to "build" to build zeebe from scratch instead of using a distball
 ARG DIST="distball"
 
-### Init image containing tini and the startup script ###
-FROM ubuntu:jammy as init
-WORKDIR /zeebe
-RUN rm /etc/apt/apt.conf.d/docker-clean
+### AMD64 base image ###
+# BASE_DIGEST_AMD64 is defined at the top of the Dockerfile
+# hadolint ignore=DL3006
+FROM ${BASE_IMAGE}@${BASE_DIGEST_AMD64} as base-amd64
+ARG BASE_DIGEST_AMD64
+ARG BASE_DIGEST="${BASE_DIGEST_AMD64}"
+
+### ARM64 base image ##
+# BASE_DIGEST_ARM64 is defined at the top of the Dockerfile
+# hadolint ignore=DL3006
+FROM ${BASE_IMAGE}@${BASE_DIGEST_ARM64} as base-arm64
+ARG BASE_DIGEST_ARM64
+ARG BASE_DIGEST="${BASE_DIGEST_ARM64}"
+
+### Base image ##
+# All package installation, updates, etc., anything with APT should be done here in a single step
+# hadolint ignore=DL3006
+FROM base-${TARGETARCH} as base
+WORKDIR /
+
+# Upgrade all outdated packages and install missing ones (e.g. locales, tini)
+# This breaks reproducibility of builds, but is acceptable to gain access to security patches faster
+# than the base image releases
+# FYI, installing packages via APT also updates the dpkg files, which are few MBs, but removing or
+# caching them could break stuff (like not knowing the package is present) or container scanners
+# hadolint ignore=DL3008
 RUN --mount=type=cache,target=/var/cache/apt,sharing=locked \
     --mount=type=cache,target=/var/lib/apt,sharing=locked \
+    --mount=type=cache,target=/var/log/apt,sharing=locked \
     apt-get -qq update && \
-    apt-get install -y --no-install-recommends tini=0.19.0-1 && \
-    cp /usr/bin/tini .
-COPY --link --chown=1000:0 docker/utils/startup.sh .
+    apt-get install -yqq --no-install-recommends tini ca-certificates && \
+    apt-get upgrade -yqq --no-install-recommends
+
+
+### AMD64 JDK image ###
+# JDK_DIGEST_AMD64 is defined at the top of the Dockerfile
+# hadolint ignore=DL3006
+FROM ${JDK_IMAGE}@${JDK_DIGEST_AMD64} as jdk-amd64
+
+### ARM64 JDK image ##
+# JDK_DIGEST_ARM64 is defined at the top of the Dockerfile
+# hadolint ignore=DL3006
+FROM ${JDK_IMAGE}@${JDK_DIGEST_ARM64} as jdk-arm64
+
+### Build custom JRE using the base JDK image
+# hadolint ignore=DL3006
+FROM jdk-${TARGETARCH} as jre-build
+
+# Build a custom JRE which will strip down and compress modules to end up with a smaller Java \
+# distribution than the official JRE. This will also include useful debugging tools like
+# jcmd, jmod, jps, etc., which take little to no space. Anecdotally, compressing modules add around
+# 10ms to the start up time, which is negligible considering our application takes ~10s to start up.
+# See https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/
+# hadolint ignore=DL3018
+RUN jlink \
+     --add-modules ALL-MODULE-PATH \
```

Comment:
:question: Could we trim that down a little bit? :pray: I.e. exclude modules such as `jdk.javadoc`, `jdk.unsupported.desktop` and probably more?

Reply:
As long as we're sure our third party code won't need it ;) But yeah, jdk.javadoc and jdk.unsupported.desktop can probably be removed.

Reply:
Excluding the following modules only trims down 3MB :shrug:

- jdk.unsupported.desktop
- jdk.desktop
- jdk.javadoc
- jdk.incubator.concurrent
- jdk.incubator.vector

If you look at the list of modules (e.g. `ls -lash $JAVA_HOME/jmods`), you'll see the largest modules are required (base, compiler, etc.). Javadoc is 1.6M, so that's nice, but the rest are kind of whatever. The XML one is surprisingly large, but needed for the `bpmn-model` stuff. So yeah, I don't think there's much to trim there unfortunately :(